File: backend/app/api/v1/investments.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps import get_current_user
from app.database import get_db
from app.models.user import User
from app.schemas.investments import (
    InvestmentCreate,
    InvestmentResponse,
    InvestmentStatusResponse,
    InvestmentSummaryResponse,
    InvestmentUpdate,
)
from app.services import investments as investments_service

router = APIRouter(prefix="/investments", tags=["investments"])


# ── Specific routes BEFORE /{investment_id} to avoid path conflicts ───────────


@router.get("/summary", response_model=InvestmentSummaryResponse)
async def get_investment_summary(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InvestmentSummaryResponse:
    """Aggregated overview of all active investments."""
    return await investments_service.get_investment_summary(db, current_user.id)


# ── Collection routes ─────────────────────────────────────────────────────────


@router.post("", response_model=InvestmentResponse, status_code=201)
async def create_investment(
    body: InvestmentCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InvestmentResponse:
    """Register a new investment (deposit, fund, stock, or bond)."""
    return await investments_service.create_investment(db, current_user.id, body)


@router.get("", response_model=list[InvestmentResponse])
async def list_investments(
    investment_type: str | None = Query(None),
    is_active: bool | None = Query(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[InvestmentResponse]:
    """List investments with optional filters."""
    return await investments_service.list_investments(
        db, current_user.id, investment_type, is_active
    )


# ── Item routes ───────────────────────────────────────────────────────────────


@router.get("/{investment_id}", response_model=InvestmentResponse)
async def get_investment(
    investment_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InvestmentResponse:
    """Get a single investment by ID."""
    return await investments_service.get_investment(db, current_user.id, investment_id)


@router.get("/{investment_id}/status", response_model=InvestmentStatusResponse)
async def get_investment_status(
    investment_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InvestmentStatusResponse:
    """Get calculated return and status for an investment."""
    return await investments_service.get_investment_status(db, current_user.id, investment_id)


@router.patch("/{investment_id}", response_model=InvestmentResponse)
async def update_investment(
    investment_id: uuid.UUID,
    body: InvestmentUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InvestmentResponse:
    """Partially update an investment."""
    return await investments_service.update_investment(db, current_user.id, investment_id, body)


@router.delete("/{investment_id}", status_code=204)
async def delete_investment(
    investment_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Delete an investment."""
    await investments_service.delete_investment(db, current_user.id, investment_id)


@router.post("/{investment_id}/renew", response_model=InvestmentResponse)
async def renew_investment(
    investment_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InvestmentResponse:
    """Manually renew a deposit or fund by extending its maturity date."""
    return await investments_service.renew_investment(db, current_user.id, investment_id)

File: frontend/src/lib/api/investments.ts
import { apiFetchJson, apiFetch } from './client';
import type {
	InvestmentResponse,
	InvestmentCreate,
	InvestmentUpdate,
	InvestmentStatusResponse,
	InvestmentSummaryResponse
} from '$lib/types';

export async function getInvestmentSummary(): Promise<InvestmentSummaryResponse> {
	return apiFetchJson<InvestmentSummaryResponse>('/api/v1/investments/summary');
}

export async function getInvestments(filters?: {
	investment_type?: string;
	is_active?: boolean;
}): Promise<InvestmentResponse[]> {
	const params = new URLSearchParams();
	if (filters?.investment_type) params.set('investment_type', filters.investment_type);
	if (filters?.is_active !== undefined) params.set('is_active', String(filters.is_active));
	const qs = params.toString();
	return apiFetchJson<InvestmentResponse[]>(`/api/v1/investments${qs ? '?' + qs : ''}`);
}

export async function getInvestment(id: string): Promise<InvestmentResponse> {
	return apiFetchJson<InvestmentResponse>(`/api/v1/investments/${id}`);
}

export async function getInvestmentStatus(id: string): Promise<InvestmentStatusResponse> {
	return apiFetchJson<InvestmentStatusResponse>(`/api/v1/investments/${id}/status`);
}

export async function createInvestment(data: InvestmentCreate): Promise<InvestmentResponse> {
	return apiFetchJson<InvestmentResponse>('/api/v1/investments', {
		method: 'POST',
		body: JSON.stringify(data)
	});
}

export async function updateInvestment(
	id: string,
	data: InvestmentUpdate
): Promise<InvestmentResponse> {
	return apiFetchJson<InvestmentResponse>(`/api/v1/investments/${id}`, {
		method: 'PATCH',
		body: JSON.stringify(data)
	});
}

export async function deleteInvestment(id: string): Promise<void> {
	const res = await apiFetch(`/api/v1/investments/${id}`, { method: 'DELETE' });
	if (!res.ok) throw new Error('Error eliminando inversión');
}

export async function renewInvestment(id: string): Promise<InvestmentResponse> {
	return apiFetchJson<InvestmentResponse>(`/api/v1/investments/${id}/renew`, {
		method: 'POST'
	});
}
